File: app-ios-nativa/Fontaneria/Fuentes/Fontaneria/Sesion.swift
import Dominio
import Foundation

/// La sesión: token de acceso en memoria, token de refresco en el llavero.
///
/// Es un actor porque la tocan a la vez la pantalla (entrar, salir) y las
/// sincronizaciones de fondo, y dos renovaciones simultáneas con el mismo
/// token de refresco harían que el servidor revocara la segunda: la rotación
/// invalida el token en cuanto se usa una vez.
public actor Sesion {
    private let cliente: ClienteApi
    private let credenciales: AlmacenCredenciales

    public private(set) var tokenAcceso: String?
    public private(set) var usuario: UsuarioApi?

    public init(cliente: ClienteApi, credenciales: AlmacenCredenciales) {
        self.cliente = cliente
        self.credenciales = credenciales
    }

    public var autenticado: Bool { tokenAcceso != nil }

    /// Cambia el código que trae el navegador por los tokens de sesión.
    public func entrar(conCodigoDeCanje codigo: String) async throws {
        try aplicar(await cliente.canjear(codigoCanje: codigo))
    }

    /// Solo contra un servidor con `PERMITIR_LOGIN_DEV=true`.
    public func entrarComoDesarrollo(email: String) async throws {
        try aplicar(await cliente.loginDeDesarrollo(email: email))
    }

    /// Recupera la sesión con el token guardado de una vez anterior.
    /// Devuelve `false` si no había, o si ya no sirve.
    public func restaurar() async -> Bool {
        guard let guardado = try? credenciales.tokenRefresco(), !guardado.isEmpty else {
            return false
        }
        do {
            try aplicar(await cliente.renovar(tokenRefresco: guardado))
            return true
        } catch {
            // Da igual por qué falló: un token que no sirve no va a servir
            // luego, y dejarlo hace que cada arranque intente renovar en
            // vano. Se borra y se entra como si no hubiera cuenta.
            try? credenciales.borrarTokenRefresco()
            tokenAcceso = nil
            usuario = nil
            return false
        }
    }

    public func cerrar() async {
        if let guardado = try? credenciales.tokenRefresco(), !guardado.isEmpty {
            // Si el servidor no contesta, la sesión se cierra aquí igual: lo
            // que no puede pasar es que el botón no haga nada.
            try? await cliente.cerrarSesion(tokenRefresco: guardado)
        }
        try? credenciales.borrarTokenRefresco()
        tokenAcceso = nil
        usuario = nil
    }

    /// Ejecuta algo que necesita el token de acceso y, si el servidor dice
    /// que ha caducado, lo renueva una vez y lo vuelve a intentar. Si la
    /// renovación también falla, el fallo se propaga: ahí ya toca volver a
    /// iniciar sesión.
    public func conReintento<T: Sendable>(
        _ peticion: @Sendable (String) async throws -> T
    ) async throws -> T {
        guard let token = tokenAcceso else {
            throw ErrorApi(mensaje: "no hay ninguna sesión iniciada", codigo: 401)
        }
        do {
            return try await peticion(token)
        } catch let fallo as ErrorApi where fallo.esSesionCaducada {
            guard await restaurar(), let renovado = tokenAcceso else {
                throw fallo
            }
            return try await peticion(renovado)
        }
    }

    private func aplicar(_ respuesta: RespuestaCanje) throws {
        tokenAcceso = respuesta.tokenAcceso
        if let recibido = respuesta.usuario {
            usuario = recibido
        }
        // La rotación obliga a guardar el token nuevo: el anterior queda
        // revocado en cuanto se usa, así que perderlo aquí deja la sesión sin
        // forma de renovarse en el siguiente arranque.
        try credenciales.guardarTokenRefresco(respuesta.tokenRefresco)
    }
}

File: app-android/fontaneria/src/main/kotlin/com/jmortizsilva/guardarenlaces/fontaneria/Sesion.kt
package com.jmortizsilva.guardarenlaces.fontaneria

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * La sesión: token de acceso en memoria, token de refresco en el almacén de credenciales.
 *
 * Dos diferencias con la de iOS, y las dos son fallos de allí que aquí no se copian (ninguno está
 * comprobado en el iPhone, están leídos en el código):
 *
 * 1. **Las renovaciones van de una en una.** Cada `POST /auth/renovar` revoca el token de refresco
 *    usado. Si dos peticiones reciben un 401 a la vez (comprobar un enlace y sincronizar al volver
 *    a la app), la segunda renovaría con un token ya revocado, fallaría, y se cerraría la sesión.
 *    En iOS la sesión es un actor, pero un actor deja entrar otra llamada mientras la primera
 *    espera a la red. Aquí hay un cerrojo, y quien llega segundo ve que el token ya cambió y usa el
 *    nuevo.
 * 2. **El token guardado solo se tira si el servidor dice 401**, que es lo que dice el contrato:
 *    «el cliente debe volver a iniciar sesión desde cero». En iOS se tira por cualquier fallo,
 *    también por no tener red: abrir la app sin conexión cerraría la sesión.
 */
class Sesion(
    private val cliente: ClienteApi,
    private val credenciales: AlmacenCredenciales,
) {
    private val cerrojo = Mutex()

    @Volatile
    var tokenAcceso: String? = null
        private set

    @Volatile
    var usuario: UsuarioApi? = null
        private set

    @Volatile private var tokenGuardado: Boolean? = null

    /**
     * Hay cuenta: hay un token de refresco guardado. Puede que todavía no haya token de acceso, por
     * ejemplo al arrancar sin red; se pedirá con la primera petición que lo necesite.
     */
    val conCuenta: Boolean
        get() = tokenGuardado ?: (credenciales.tokenRefresco() != null).also { tokenGuardado = it }

    sealed interface Restauracion {
        data object SinCuenta : Restauracion

        data object Restaurada : Restauracion

        /** Hay cuenta, pero ahora no se puede hablar con el servidor. La cuenta se conserva. */
        data class SinConexion(val fallo: ErrorApi) : Restauracion
    }

    /** Cambia el código que trae el navegador por los tokens de sesión. */
    suspend fun entrar(codigoDeCanje: String) = cerrojo.withLock {
        aplicar(cliente.canjear(codigoDeCanje))
    }

    /** Solo contra un servidor con `PERMITIR_LOGIN_DEV=true`. */
    suspend fun entrarComoDesarrollo(email: String) = cerrojo.withLock {
        aplicar(cliente.loginDeDesarrollo(email))
    }

    /** Recupera la sesión con el token guardado de una vez anterior. */
    suspend fun restaurar(): Restauracion {
        if (!conCuenta) return Restauracion.SinCuenta
        return try {
            renovar(caducado = null)
            Restauracion.Restaurada
        } catch (fallo: ErrorApi) {
            if (conCuenta) Restauracion.SinConexion(fallo) else Restauracion.SinCuenta
        }
    }

    suspend fun cerrar() = cerrojo.withLock {
        credenciales.tokenRefresco()?.let { guardado ->
            // Si el servidor no contesta, la sesión se cierra aquí igual: lo que no puede pasar
            // es que el botón no haga nada.
            try {
                cliente.cerrarSesion(guardado)
            } catch (_: ErrorApi) {}
        }
        olvidar()
    }

    /**
     * Ejecuta algo que necesita el token de acceso. Si no lo hay todavía, lo pide; si el servidor
     * dice que ha caducado, lo renueva una vez y lo vuelve a intentar. Si la renovación también
     * falla, el fallo se propaga: ahí ya toca volver a iniciar sesión.
     */
    suspend fun <T> conReintento(peticion: suspend (String) -> T): T {
        val token = tokenAcceso ?: renovar(caducado = null)
        return try {
            peticion(token)
        } catch (fallo: ErrorApi) {
            if (!fallo.esSesionCaducada) throw fallo
            peticion(renovar(caducado = token))
        }
    }

    /**
     * Un token de acceso nuevo, pidiéndolo solo si hace falta: si mientras se esperaba el cerrojo
     * otra llamada ya lo renovó, se usa ese.
     */
    private suspend fun renovar(caducado: String?): String = cerrojo.withLock {
        tokenAcceso
            ?.takeIf { it != caducado }
            ?.let { yaRenovado ->
                return yaRenovado
            }

        val guardado =
            credenciales.tokenRefresco() ?: throw ErrorApi("no hay ninguna sesión iniciada", 401)
        try {
            aplicar(cliente.renovar(guardado))
        } catch (fallo: ErrorApi) {
            if (fallo.esSesionCaducada) olvidar()
            throw fallo
        }
        checkNotNull(tokenAcceso)
    }

    private fun aplicar(respuesta: RespuestaCanje) {
        // La rotación obliga a guardar el token nuevo antes que nada: el anterior queda revocado
        // en cuanto se usa, así que perder este deja la sesión sin forma de renovarse.
        credenciales.guardarTokenRefresco(respuesta.tokenRefresco)
        tokenGuardado = true
        tokenAcceso = respuesta.tokenAcceso
        respuesta.usuario?.let { usuario = it }
    }

    private fun olvidar() {
        credenciales.borrarTokenRefresco()
        tokenGuardado = false
        tokenAcceso = null
        usuario = null
    }
}
